File: readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

bool isFinalPunctuation(char letter);

int main(void)
{
    string userInput = get_string("Text: ");
    // get user entry length
    int length = strlen(userInput);

    // letters to be counted
    int letters = 0;
    // spaces are used to count words
    int words = 1;
    // punctuation signs used to count sentences
    int sentences = 0;

    // iterate over each character in the user entry
    for (int i = 0; i < length; i++)
    {
        // char is alphanumeric
        if (isalpha(userInput[i]))
        {
            letters++;
        }
        // char is space
        else if (isspace(userInput[i]))
        {
            words++;
        }
        // char is punctuation that split sentence
        else if (isFinalPunctuation(userInput[i]) && !isFinalPunctuation(userInput[i - 1]))
        {
            sentences++;
        }
    }

    float grade = 0.0588 * (100 * (float) letters / (float) words) - 0.296 * (100 * (float) sentences / (float) words) - 15.8;

    // check grade to print the expected result
    if (grade < 16 && grade > 1)
    {
        printf("Grade %i\n", (int) round(grade));
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Before Grade 1\n");
    }
}

bool isFinalPunctuation(char letter)
{
    return letter == '!' || letter == '.' || letter == '?';
}

File: readability.py
from cs50 import get_string

# main function
def main():
    user_input = get_string("Text: ")
    # get user entry length
    length = len(user_input)

    # letters to be counted
    letters = 0
    # spaces are used to count words
    words = 1
    # punctuation signs used to count sentences
    sentences = 0

    # iterate over each character in the user entry
    for i in range(length):
        # counts letters using ascii code
        if (ord(user_input[i]) >= 65 and ord(user_input[i]) <= 122):
            letters += 1

        # counts words by reading spaces
        elif (user_input[i].isspace()) == True:
            words += 1

        # counts sentences by finding dots, exclamation marks and interrogatives
        elif isFinalPunctuation(user_input[i]) and not isFinalPunctuation(user_input[i - 1]):
            sentences += 1

    print(f"letters {letters}")
    print(f"sentences {sentences}")
    print(f"words {words}")

    L = letters * 100 / words
    S = sentences * 100 / words
    # calculating Coleman-Liau index
    index = round(0.0588 * L - 0.296 * S - 15.8)

    # check grade to print the expected result
    if (index < 1):
        print("Before Grade 1")
    elif (index >= 16):
        print("Grade 16+")
    else:
        print(f"Grade {index}")

# method to check final punctuations characters
def isFinalPunctuation(letter):
    return ord(letter) == 33 or ord(letter) == 46 or ord(letter) == 63

main()
